src/expr/lazy_proof.cpp

```diff
@@ -0,0 +1,176 @@
+/*********************                                                        */
+/*! \file lazy_proof.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Lazy proof utility
+ **/
+
+#include "expr/lazy_proof.h"
+
+using namespace CVC4::kind;
+
+namespace CVC4 {
+
+LazyCDProof::LazyCDProof(ProofNodeManager* pnm,
+                         ProofGenerator* dpg,
+                         context::Context* c)
+    : CDProof(pnm, c), d_gens(c ? c : &d_context), d_defaultGen(dpg)
+{
+}
+
+LazyCDProof::~LazyCDProof() {}
+
+std::shared_ptr<ProofNode> LazyCDProof::mkProof(Node fact)
+{
+  Trace("lazy-cdproof") << "LazyCDProof::mkLazyProof " << fact << std::endl;
+  // make the proof, which should always be non-null, since we construct an
+  // assumption in the worst case.
+  std::shared_ptr<ProofNode> opf = CDProof::mkProof(fact);
+  Assert(opf != nullptr);
+  if (!hasGenerators())
+  {
+    Trace("lazy-cdproof") << "...no generators, finished" << std::endl;
+    // optimization: no generators, we are done
+    return opf;
+  }
+  // otherwise, we traverse the proof opf and fill in the ASSUME leafs that
+  // have generators
```

Comment:
Please also say why you need to do a custom traversal and not use something like `getFreeAssumptions` and fill the generators of those guys.

Reply:
There isn't anything custom here.  In theory this traversal could be refactored to use a common utility as getFreeAssumptions.  Although note that getFreeAssumptions will be later refactored as well (it should not live in ProofNode).  This is future work.

Reply:
I'd prefer having this clear by using `getFreeAssumptions` now. It'd just make this method clear now and later when getFreeAssumptions changes to some util it's easy to update this method.

Reply:
`getFreeAssumptions` is not in master, not is it going to be merged to master. This will be refactored later to use a proof traversal utility, or be internal to ProofNodeManager.

Reply:
Sorry, I meant to say "getFreeAssumptionsMap", which is what would be required here.